pkg/controller/certificates/internal/secretsmanager/secret.go

```diff
@@ -205,6 +205,7 @@ func (s *SecretsManager) setValues(crt *cmapi.Certificate, secret *corev1.Secret
 	secret.Annotations[cmapi.CertificateNameKey] = crt.Name
 	secret.Annotations[cmapi.IssuerNameAnnotationKey] = crt.Spec.IssuerRef.Name
 	secret.Annotations[cmapi.IssuerKindAnnotationKey] = apiutil.IssuerKind(crt.Spec.IssuerRef)
+	secret.Annotations[cmapi.IssuerGroupAnnotationKey] = crt.Spec.IssuerRef.Group
```

Comment:
This will overwrite any existing group annotation key, but what about people who have already deployed the buggy cert-manager-0.16....they will still have secrets without the issuerRef.group set. Does that mean that they will have to manually set this annotation or delete the secret and start again? Do we need to add a release note explaining how to fix broken secrets? 

If user modifies `Certificate.Spec.IssuerRef.Group` the certificate will be immediately renewed, even though the certificate may still be valid. Is that ok?

Reply:
This has no effect on current cert-manager users not using external issuers. The issue was the fact that this was not set, so defaulting to the cert-manager group, which did not equal the external issuer group. Here is the trigger controller policy 

```
func SecretHasUpToDateIssuerAnnotations(input Input) (string, string, bool) {
	name := input.Secret.Annotations[cmapi.IssuerNameAnnotationKey]
	kind := input.Secret.Annotations[cmapi.IssuerKindAnnotationKey]
	group := input.Secret.Annotations[cmapi.IssuerGroupAnnotationKey]
	if name != input.Certificate.Spec.IssuerRef.Name ||
		!issuerKindsEqual(kind, input.Certificate.Spec.IssuerRef.Kind) ||
		!issuerGroupsEqual(group, input.Certificate.Spec.IssuerRef.Group) {
		return "IncorrectIssuer", fmt.Sprintf("Issuing certificate as Secret was previously issued by %s", formatIssuerRef(name, kind, group)), true
	}
	return "", "", false
}
```


```
func issuerGroupsEqual(l, r string) bool {
	if l == "" {
		l = defaultIssuerGroup
	}
	if r == "" {
		r = defaultIssuerGroup
	}
	return l == r
}
```

Upgrading cert-manager shouldn't cause a re-issuance for all users, only those using external issuers, which is what the bug is anyway :sweat_smile: 

> If user modifies Certificate.Spec.IssuerRef.Group the certificate will be immediately renewed, even though the certificate may still be valid. Is that ok?

That is exactly the behavior we want. If a user changes the issuer ref group, they are now requesting that the certificate be signed by a completely different issuer, which should always cause a re-issuance.

Reply:
Just to be a bit more clear:

## 0.16.0
internal issuer -> No annotation -> **accept** cert.IssuerRef.Group = cert-manager.io
external issuer -> No annotation -> **reject** cert.IssuerRef.Group = external.issuer.io

## 0.16.1
internal issuer -> issuer annotation cert-manager.io -> **accept** cert.IssuerRef.Group = cert-manager.io
external issuer -> issuer annotation external.issuer.io -> **accept** cert.IssuerRef.Group = external.issuer.io